clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -0,0 +1,190 @@
+import org.apache.kafka.common.network.NetworkSend;
+import org.apache.kafka.common.network.NetworkReceive;
+import org.apache.kafka.common.network.TransportLayer;
+import org.apache.kafka.common.security.auth.PrincipalBuilder;
+
+
+public class SaslServerAuthenticator implements Authenticator {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SaslServerAuthenticator.class);
+    private SaslServer saslServer;
+    private Subject subject;
+    private TransportLayer transportLayer;
+    private String node = "0";
+    private NetworkReceive netInBuffer;
+    private NetworkSend netOutBuffer;
+    private SaslServerCallbackHandler saslServerCallbackHandler;
+    private PrincipalBuilder principalBuilder;
+
+    public SaslServerAuthenticator(String node, final Subject subject) throws IOException {
+        this.node = node;
+        this.transportLayer = transportLayer;
+        this.subject = subject;
+        saslServer = createSaslServer();
+    }
+
+    public void configure(TransportLayer transportLayer, PrincipalBuilder principalBuilder) {
+        this.transportLayer = transportLayer;
+        this.principalBuilder = principalBuilder;
+    }
+
+    private SaslServer createSaslServer() throws IOException {
+        if (subject != null) {
+            // server is using a JAAS-authenticated subject: determine service principal name and hostname from kafka server's subject.
+            if (subject.getPrincipals().size() > 0) {
+                try {
+                    saslServerCallbackHandler = new SaslServerCallbackHandler(Configuration.getConfiguration());
+                    final Object[] principals = subject.getPrincipals().toArray();
+                    final Principal servicePrincipal = (Principal) principals[0];
+
+                    final String servicePrincipalNameAndHostname = servicePrincipal.getName();
+                    int indexOf = servicePrincipalNameAndHostname.indexOf("/");
+                    final String servicePrincipalName = servicePrincipalNameAndHostname.substring(0, indexOf);
+                    final String serviceHostnameAndKerbDomain = servicePrincipalNameAndHostname.substring(indexOf + 1, servicePrincipalNameAndHostname.length());
+                    indexOf = serviceHostnameAndKerbDomain.indexOf("@");
+                    final String serviceHostname = serviceHostnameAndKerbDomain.substring(0, indexOf);
+                    final String mech = "GSSAPI";
+
+                    LOG.debug("serviceHostname is '" + serviceHostname + "'");
+                    LOG.debug("servicePrincipalName is '" + servicePrincipalName + "'");
+                    LOG.debug("SASL mechanism(mech) is '" + mech + "'");
+                    boolean usingNativeJgss = Boolean.getBoolean("sun.security.jgss.native");
+                    if (usingNativeJgss) {
+                        try {
+                            GSSManager manager = GSSManager.getInstance();
```

Comment:
I am guessing this is all part of GSS API magic but a link to doc or some explanation on what we are doing here might help with future maintenance.